docs/design-docs/01-oci-support.md

```diff
@@ -0,0 +1,418 @@
+config as image content. For security and production configuration management, if a 
+company has practices for managing Docker container images in private registries, then
+the same practices and security model can be applied to config package images in private
+registries.
+
+https://github.com/GoogleContainerTools/kpt/issues/2300
+
+## Design
+
+### Design Assumptions
+
+The first stage of OCI support comes from adding support for `oci` in the places
+where `git` appears today.
+
+An image tag is used in the same way a git branch or tag would be used.
+
+An image digest is used in the same way a git commit would be used.
+
+The scope of a single image is one root package, with any number of optional sub-packages.
+
+The structure of the image is a single tar layer. The root `Kptfile` is in the base directory from the tar layer's point of view. It contains only the Kpt package files, no entrypoint or executables.
+
+A package image should not be confused with a container image. Container images are executable by software, like `Docker`, and package images are purely configuration data.
+
+### Config chages
+
+The `Kptfile` structures for `upstream` and `upstreamLock` have `oci` in addition to `git` properties. The `type` property also has the string `oci` added as an accepted value.
+
+```yaml
+upstream:
+  type: oci
+  oci:
+    image: 'IMAGE:TAG'
+upstreamLock:
+  type: oci
+  oci:
+    image: 'IMAGE:DIGEST'
+```
+
+New verions of `kpt` will support existing `Kptfile`. The file structures and `git` functionality is unchanged.
+
+Existing versions of `kpt` will support `Kptfile` with `upstream` based on `git` for the same reason. The structure and meaning of existing fields is not changed.
+
+Existing versions of `kpt` will not support `Kptfile` with `upstream` based on `oci`. The `type` value, and missing `git` information will fail validation. The `kpt` binary used will need to be upgraded.
+
+### Command changes
+
+### `kpt pkg get`
+
+The argument that determines upstream today is parsed into `repo`, `ref`, and `path`, and is implicitly a `git` location.
+
+To support `oci`, it will be necessary to extract different values in a way that's unambiguous. Unfortunately, OCI image names have no Uri prefix, and are indistinguishable from a valid path or file name.
+
+To solve this, using [Helm](https://helm.sh/docs/topics/registries/#other-subcommands) as an example, the prefix `oci://` can be used. This ensures that selecting `oci` protocol isn't accidental, and it won't collide with other location formats that may be added.
+
+```shell
+# clone package as new folder
+kpt pkg get oci://us-docker.pkg.dev/the-project-id/the-repo-name/the-package:v3 my-package
+```
+
+Because OCI image reference already has a convention for `image:tag` references, using `:v3` should be used instead of `@v3` for version. It will be more intuitive how it relates to the registry, and easier to cut and paste values.
+
+### `kpt pkg get` sub-packages
```

Comment:
There are 2 distinct scenarios:
1. Building up larger packages from smaller nested subpackages
2. Extracting pieces from larger packages.

The first is more important than the second.

Reply:
Agreed that addressing a path in a package is mainly (only?) to support sub-package upstreams.

As far as single-root-package vs multi-root-package images goes, I'd suggest as a rule that an OCI image would require a Kptfile in the root directory and that is the only one that would have `kpt live` behavior.

Reason being (thinking of helm charts, npm packages, etc) putting 2 top-level units one package makes the use of name and version of the package ambiguous.

In the case of a literal mapping a git mono-repo with per-folder package to OCI, I believe that would be mono-registry with per-image packages.

Reply:
Resolving thread - in the prototype code a double-slash can add an upstream sub-package `kpt pkg get oci://my-registry,io/my-image//my/path:v5`.

The sub-path isn't supported on pull and push at the moment. Push in particular would require more work (needing to know where to get the rest of the package from - otherwise it'll push only that single subfolder.)

Also because pull/push acts on one package at a time - there isn't that much of a saving extracting only one folder to change and push back...